blog/tech-talk-window-resize-behavior.md

```diff
@@ -0,0 +1,290 @@
+- [`IDCompositionDevice::Commit`](https://source.chromium.org/chromium/chromium/src/+/main:ui/gl/dc_layer_tree.cc;l=1034;drc=35d26d364efb57d0386b98312ba739f7f65ae97e) — This updates the clip rect.
+
+Here's what's important to understand: both functions return synchronously on the CPU. However, they schedule tasks that run asynchronously on the GPU at a later time. Windows and its services (such as [DWM](https://en.wikipedia.org/wiki/Desktop_Window_Manager)) decide when they will run and in which order. So they take effect asynchronously, and not always within the same frame.
+
+Unfortunately, Windows provides no way for us to synchronize those operations. So I had to find other approaches to fix this.
+
+There were two options that I evaluated with the Chromium maintainers:
+
+1. While resizing, paint all previously drawn areas outside of the viewport transparent. This makes those areas invisible. It fixes the artifacts.
+2. While resizing, switch from an `IDXGISwapChain1` to a DirectComposition surface which synchronizes updates with `IDCompositionDevice::Commit`. This also fixes the artifacts.
+
+We went with the first option because it leads to faster resizes than the second option.
+
+I landed a [patch](https://crrev.com/c/7115438) in Chromium that implemented that first solution.
+
+I also submitted two other patches in preparation for the main patch:
+
+1. The [first one](https://crrev.com/c/7129658) fixed a bug in existing code that would make CI fail in combination with the main patch. It also made launching Electron apps and Chrome a tiny bit faster.
+2. The [second one](https://crrev.com/c/7210913) was split off to make code review easier on the main patch.
+
+## Second bug
+
+In addition to this first bug, there was a second bug which also led to stale pixels.
+
+Here's what was going on there:
+
+When the user resizes the window, Chromium needs to redraw the contents of the window for the new window size. This takes some time. The new frame isn't ready immediately.
+
+Here's a sequence of frames that demonstrates this:
+
+![First frame of the Chromium resize sequence](/assets/img/blog/tech-talk-window-resize-behavior/chromium-resize-sequence-1.png)
+![Second frame of the Chromium resize sequence](/assets/img/blog/tech-talk-window-resize-behavior/chromium-resize-sequence-2.png)
+![Third frame of the Chromium resize sequence](/assets/img/blog/tech-talk-window-resize-behavior/chromium-resize-sequence-3.png)
+![Fourth frame of the Chromium resize sequence](/assets/img/blog/tech-talk-window-resize-behavior/chromium-resize-sequence-4.png)
+![Fifth frame of the Chromium resize sequence](/assets/img/blog/tech-talk-window-resize-behavior/chromium-resize-sequence-5.png)
```

Comment:
Should we splice this into a `GIF` with slower timings between the frames?

Reply:
We could. This is something that I didn't do for time reasons.